src/ripple/test/jtx/basic_prop.h

```diff
@@ -0,0 +1,86 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_TEST_JTX_BASIC_PROP_H_INCLUDED
+#define RIPPLE_TEST_JTX_BASIC_PROP_H_INCLUDED
+
+namespace ripple {
+namespace test {
+namespace jtx {
+
+struct basic_prop
+{
+    virtual ~basic_prop() = default;
+    virtual std::unique_ptr<basic_prop> clone() const = 0;
+};
+
+namespace detail {
+
+struct prop_list
+{
+    prop_list() = default;
+
+    prop_list(prop_list const& src)
+    {
+        for (auto& prop : src.list)
+        {
+            list.emplace_back(prop->clone());
+        }
+    }
+
+    prop_list& operator=(prop_list const& src)
+    {
+        list.clear();
+        for (auto& prop : src.list)
+        {
+            list.emplace_back(prop->clone());
+        }
+        return *this;
+    }
+
+    std::vector<std::unique_ptr<
+        basic_prop>> list;
+};
+
+template <class T>
```

Comment:
Very nice job here. Just a note that these class are doing type erasure. A type erasure library that did small buffer optimization would be a win here (if we could just plug it in). Unfortunately, I couldn't find one in boost (Apparently `boost::spirit::hold_any` is such a beast, but it requires `operator<<` be defined, and it seems like a hack). I'm making this note in case you or anyone else here knows of a good type erasure library that does SBO.